flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/utils/ValuesOperationFactory.java

```diff
@@ -0,0 +1,201 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.table.expressions.ApiExpressionUtils.valueLiteral;
+
+/**
+ * Utility class for creating valid {@link ValuesQueryOperation} operation.
+ */
+@Internal
+class ValuesOperationFactory {
+	/**
+	 * Creates a valid {@link ValuesQueryOperation} operation.
+	 *
+	 * <p>It derives a row type based on {@link LogicalTypeGeneralization}. It flattens any
+	 * row constructors. It does not flatten ROWs which are a result of e.g. a function call.
+	 *
+	 * <p>The resulting schema can be provided manually. If it is not, the schema will be automatically derived from
+	 * the types of the expressions.
+	 */
+	QueryOperation create(
+			@Nullable TableSchema expectedSchema,
+			List<ResolvedExpression> resolvedExpressions,
+			ExpressionResolver.PostResolverFactory postResolverFactory) {
+		List<List<ResolvedExpression>> resolvedRows = unwrapFromRowConstructor(resolvedExpressions);
+
+		TableSchema schema = Optional.ofNullable(expectedSchema)
+			.orElseGet(() -> extractSchema(resolvedRows));
+
+		List<List<ResolvedExpression>> castedExpressions = resolvedRows.stream()
+			.map(row -> convertToExpectedRowType(postResolverFactory, schema.getFieldDataTypes(), row))
+			.collect(Collectors.toList());
+
+		return new ValuesQueryOperation(castedExpressions, schema);
+	}
+
+	private TableSchema extractSchema(List<List<ResolvedExpression>> resolvedRows) {
+		DataType[] dataTypes = findRowType(resolvedRows);
+		String[] fieldNames = IntStream.range(0, dataTypes.length)
+			.mapToObj(i -> "f" + i)
+			.toArray(String[]::new);
+		return TableSchema.builder()
+			.fields(fieldNames, dataTypes)
+			.build();
+	}
+
+	private List<ResolvedExpression> convertToExpectedRowType(
+			ExpressionResolver.PostResolverFactory postResolverFactory,
+			DataType[] dataTypes,
+			List<ResolvedExpression> row) {
+		return IntStream.range(0, row.size())
+			.mapToObj(i -> {
+				boolean typesMatch = row.get(i)
+					.getOutputDataType()
+					.getLogicalType()
+					.equals(dataTypes[i].getLogicalType());
+				if (typesMatch) {
+					return row.get(i);
+				}
+
+				ResolvedExpression castedExpr = row.get(i);
+				DataType targetType = dataTypes[i];
+				return convertToExpectedType(castedExpr, targetType, postResolverFactory);
+			})
+			.collect(Collectors.toList());
+	}
+
+	private ResolvedExpression convertToExpectedType(
+			ResolvedExpression castedExpr,
+			DataType targetType,
+			ExpressionResolver.PostResolverFactory postResolverFactory) {
+
+		// if the expression is a literal try converting the literal in place instead of casting
+		if (castedExpr instanceof ValueLiteralExpression) {
+			Optional<?> convertedValue = ((ValueLiteralExpression) castedExpr).getValueAs(targetType.getConversionClass());
+			if (convertedValue.isPresent()) {
+				return valueLiteral(convertedValue.get(), targetType);
+			}
+		}
+		return postResolverFactory.cast(castedExpr, targetType);
```

Comment:
Should we add an additional check if the expression can be casted to the target type? The post resolver factory should be called with care because no validation is performed anymore when the schema is manually defined.

Reply:
Yes, forgot about this after adding the method with a user provided type. It was unnecessary when the common type was derived automatically.